### ISSUE TITLE
Allow witnesses to report their head block at any time

### ISSUE THREAD

**Issue by bytemaster** [steemit/steem#319]
This will advance the last irreversible block faster and enable better performance on subchains.

**Comment by theoreticalbts** [comment]
These messages will spam the blockchain.  We're currently operating at ~5 tx / block (actually I took a sample of a few blocks and saw an average of 3.65 tx / block).  If every witness adds a transaction to every block, it'll bloat our transaction load by a factor of 5.  I don't know of the p2p code will be able to handle it.

Also, there are dragons in precisely defining the semantics of reports.  I'm thinking that when witness W reports block B, it is promising not to sign any chain with a different block B' at the same height as B.  Suppose that after B, the other witnesses all switch to a different chain.  W cannot switch without violating his promise; but if W doesn't switch, then he will never participate in the main chain and forever remain on a minority fork.

Some kind of weaker promise may be necessary, but if the promise is too weak then it turns into purely advisory information which can't be used as a primitive (e.g. for providing atomicity with side-chains).

**Comment by theoreticalbts** [comment]
This may actually be doable. Currently, when signing a block, each witness is implicitly promising they won't sign any other chain unless it is a majority fork. The semantics of the HBR (head block report) should be similar -- the witness won't sign any chain not based on the HBR unless it is a majority fork.

Once enough negative promises have accumulated, we know that no majority fork will be able to exist (provided everyone keeps their promises), and can advance LIB. The `STEEMIT_IRREVERSIBLE_THRESHOLD` for advancement of LIB is basically a buffer against witnesses who break this promise.

**Comment by theoreticalbts** [comment]
Irreversibility is a non-consensus item. Therefore, we can put head block reports on a sidechain -- or for that matter, a single centralized server! Anyone can verify HBR's against a witness's current block signing key. Therefore a server with private upstream to witnesses, but public downstream to clients, is the ideal solution.

Basically I'm thinking:

- The server will be a callback API which streams head block ID and signature to subscribers
- The client will be a plugin which connects to the server and subscribes to the remote callback API

This much is straightforward and allows a client to receive authenticated HBID updates by connecting directly to a witness.

Now all we need is a simple JS / Python script (or any language with websocket support) which allows us to reverse proxy this connection, so instead of everyone who wants streaming HBID updates connecting directly to a witness (opening the door to DDOS attacks on witnesses by exposing their IP addresses), instead a secondary server controlled by the witness connects to the witness node, and everyone else connects to the secondary server. The secondary server address can be published as metadata, or alternatively its connection info may be distributed privately to aggregators who operate public nodes.

Basically we don't want HBID updates to go over the main p2p network because they're spammy non-consensus data which most nodes don't care about.

This semi-centralized architecture is basically a p2p network without peer discovery. Maybe we should just skip this step and make it a sidechain? It is interesting because it is a very minimalist sidechain with a useful practical application.

**Comment by bytemaster** [comment]
They are part of consensus because it will impact fork resolution and undo history.

**Comment by bytemaster** [comment]
Furthermore, the purpose of this "consensus" change is to enable subchains and side chains to RELY on the information and the commitment to non-reversal.

**Comment by bytemaster** [comment]
This issue changes the semantics for when witnesses should switch forks and therefore has a much bigger impact on the consensus process than simply advancing the last irreversible block.  

Under traditional DPOS each block is a vote for a fork and witnesses switch to the longest chain which is the chain with the most votes.  Due to the enforced time delays on voting, the majority can always produce blocks faster than the minority and therefore can eventually overcome any minority chain.  

The last irreversible block (LRB) concept is when there is cryptographic proof that the majority of active witnesses are on a particular fork.  Once that proof exists there is no possibility of any other fork prior to LRB unless collusion and/or double production by majority of witnesses.

Once witnesses can "vote" out of turn, the concept of consensus is completely different because a shorter chain can have "more votes".  

This means that each chain has a "virtual length" of LRB + unique votes and that witnesses should only switch forks if the "virtual length" of another fork is longer than the "virtual length" of your current fork.  This means knowing the "virtual length" of each fork.  

The virtual length can be calculated with a specialized transaction visitor that iterates over all transactions in a block and verifies virtual length. This could be error prone given changing votes, permissions, and witness rounds.

Each report counts as a vote for all blocks prior to the report.

Needless to say, this ticket needs MUCH more thought and testing before it can be considered for a production chain.

**Comment by mvandeberg** [comment]
Closing this as the idea is too abstract at the moment. If a feature such as this is required in the future, a new ticket will be created for it.
